rbac/management/role/relation_api_dual_write_handler.py

```diff
@@ -34,16 +38,116 @@ class DualWriteException(Exception):
     pass
 
 
+class ReplicationEventType(str, Enum):
+    """Replication event type."""
+
+    CREATE_SYSTEM_ROLE = "create_system_role"
+    UPDATE_SYSTEM_ROLE = "update_system_role"
+    DELETE_SYSTEM_ROLE = "delete_system_role"
+    CREATE_CUSTOM_ROLE = "create_custom_role"
+    UPDATE_CUSTOM_ROLE = "update_custom_role"
+    DELETE_CUSTOM_ROLE = "delete_custom_role"
+    ASSIGN_ROLE = "assign_role"
+    UNASSIGN_ROLE = "unassign_role"
+    CREATE_GROUP = "create_group"
+    UPDATE_GROUP = "update_group"
+    DELETE_GROUP = "delete_group"
+
+
+class ReplicationEvent:
+    """What tuples changes to replicate."""
+
+    type: ReplicationEventType
+    partition_key: str
+    add: list[common_pb2.Relationship]
+    remove: list[common_pb2.Relationship]
+
+    def __init__(
+        self,
+        type: ReplicationEventType,
+        partition_key: str,
+        add: list[common_pb2.Relationship] = [],
+        remove: list[common_pb2.Relationship] = [],
+    ):
+        """Initialize ReplicationEvent."""
+        self.partition_key = partition_key
+        self.type = type
+        self.add = add
+        self.remove = remove
+
+
+class RelationReplicator(ABC):
+    """Type responsible for replicating relations to Kessel Relations."""
+
+    @abstractmethod
+    def replicate(self, event: ReplicationEvent):
+        """Replicate the given event to Kessel Relations."""
+        pass
+
+
+class OutboxReplicator(RelationReplicator):
+    """Replicates relations via the outbox table."""
+
+    def __init__(self, role):
+        """Initialize OutboxReplicater."""
+        self.role = role
+
+    def replicate(self, event: ReplicationEvent):
+        """Replicate the given event to Kessel Relations via the Outbox."""
+        payload = self._build_replication_event(event.add, event.remove)
+        self._save_replication_event(payload, event.type, event.partition_key)
+
+    def _build_replication_event(self, relations_to_add, relations_to_remove):
+        """Build replication event."""
+        logger.info("[Dual Write] Build Replication event for role(%s): '%s'", self.role.uuid, self.role.name)
+        add_json = []
+        for relation in relations_to_add:
+            add_json.append(relationship_to_json(relation))
```

Comment:
namespace is missing in generated event (relationship_to_json needs to be updated):
```
{
  "relations_to_add": [
    {
      "resource": {
        "type": "role_binding",
        "id": "088a5150-b8ee-4ed0-868c-ca3d1417837e"
      },
      "relation": "granted",
      "subject": {
        "type": "role",
        "id": "527cf63b-69a8-4ca9-afb3-21d5d625b18b"
      }
    },
    {
      "resource": {
        "type": "role",
        "id": "527cf63b-69a8-4ca9-afb3-21d5d625b18b"
      },
      "relation": "inventory_hosts_write",
      "subject": {
        "type": "user",
        "id": "*"
      }
    },
    {
      "resource": {
        "type": "workspace",
        "id": "123456"
      },
      "relation": "parent",
      "subject": {
        "type": "workspace",
        "id": "13431712"
      }
    },
    {
      "resource": {
        "type": "workspace",
        "id": "123456"
      },
      "relation": "user_grant",
      "subject": {
        "type": "role_binding",
        "id": "088a5150-b8ee-4ed0-868c-ca3d1417837e"
      }
    }
  ],
  "relations_to_remove": []
}
```

Reply:
I addressed this by trying out using the proto schema directly and it seems to work pretty well!